fix: issues for anthropic, refactor BaseReader.ts

tanevanwifferen opened the pull request:
There were some bugs where anthropic responses would only show up in parts (parts were missing) and also tool arguments weren't handled correctly. 

Also some bugs where message.1.content.split() is not a function, because it was an object instead of a string

nanbingxyz commented on src/intellichat/readers/BaseReader.ts:
```diff
@@ -28,105 +93,174 @@ export default abstract class BaseReader implements IChatReader {
+    };
 
     try {
-      while (!done) {
```
This part also handles a scenario where some service providers or intermediaries may return content that is not always complete messages, but rather multiple incomplete messages. Therefore, if JSON parsing fails, it needs to attempt to concatenate the subsequent returned content and parse it again. MAX_SPLICE = 3 means that it will try to concatenate up to 3 times (which can cover the vast majority of cases).

e.g.

`data: {"choices":[{"content_filter_results":{},"delta":{"content":null,"role":"assistant","tool_calls":[{"function":{"arguments":"","name":"sea`

`rch_notes"},"id":"call_iXWEt3RGcPnNzT8Zn52HRTVj","index":0,"type":"function"}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1734063838,"id":"chatcmpl-AdrVWRCpJ0jAsXlclXQGwLVyyiPbp","model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_8`

`4ac7e2412"}data: {"choices":[{"content_filter_results":{},"delta":{"tool_calls":[{"function":{"arguments":"{\\""},"index":0}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1734063838,"id":"chatcmpl-AdrVWRCpJ0jAsXlclXQGwLVyyiPbp","model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_84ac7e2412"}`

tanevanwifferen replied:
which providers make use of this? I've got a fix in the pipe, but it requires each provider to override the shouldCombine and getCombine methods. 

Anthropic sends weird messages, where each tool_call message is a JSON object, but looks like below. So the smartest is to leave combining implementation to the subclass

```
event: content_block_start
data: {"type":"content_block_start","index":1,"content_block":{"type":"tool_use","id":"toolu_01DAMWvBc3e5hxwq3gu6Ec7y","name":"mcp-obsidian--list_files_in_dir","input":{}}          }

event: content_block_delta
data: {"type":"content_block_delta","index":1,"delta":{"type":"input_json_delta","partial_json":""}           }

event: content_block_delta
data: {"type":"content_block_delta","index":1,"delta":{"type":"input_json_delta","partial_json":"{\"dirp"}              }

event: content_block_delta
data: {"type":"content_block_delta","index":1,"delta":{"type":"input_json_delta","partial_json":"ath\": \""} }

event: content_block_delta
data: {"type":"content_block_delta","index":1,"delta":{"type":"input_json_delta","partial_json":"togoder"}  }

event: content_block_delta
data: {"type":"content_block_delta","index":1,"delta":{"type":"input_json_delta","partial_json":"\"}"}        }

event: content_block_stop
data: {"type":"content_block_stop","index":1       }
```

tanevanwifferen replied:
See this commit 

https://github.com/nanbingxyz/5ire/commit/9e871124bd90504cfb0fee4eff2a2442a30cb8b4

nanbingxyz replied:
Mostly some self-hosted intermediary services, which likely arises because the server hasn’t been timely flushed. Currently, many users have set up their own relay services. These services might be backed by any major model providers like OpenAI, Anthropic, Google, etc. Therefore, the possibility of outputs being incomplete JSON must be taken into account. This is a bothersome but common scenario.

tanevanwifferen replied:
Also, anthropic likes to put "ping" messages somewhere halfway the response, so we can't assume all messages are related. I would truly advise to put message combining in the subclass readers



event: content_block_start
data: {"type":"content_block_start","index":0,"content_block":{"type":"text","text":""}      }

event: ping
data: {"type": "ping"}

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":"\n\nAh, I found it"} }


I guess I could have both. One that looks if it can combine JSON when it's not combined, but put another combine logic in the concrete readers for the partial_json stuff that anthropic sends

tanevanwifferen replied:
should be fixed now, can you review again? It checks if a chunk can be JSON parsed, and if not it will try to combine it with the next chunks until it can parse the message as JSON

noxy-k commented:
@tanevanwifferen are you re-submitting with that added safety mechanism?

tanevanwifferen commented:
closes #26 #27

nanbingxyz reviewed (changes requested):
Running `npm run test -- reader.spec.ts` fails, and it seems there are some issues that need to be addressed.

tanevanwifferen commented:
I've fixed part of the unit tests, and changed the unit tests too, to match with the spec of [server side events](https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events). The JSON objects can be sent partially per line, but each line should start with data:

nanbingxyz commented:
Thanks, it's so cool!

lnvestor commented:
@nanbingxyz @tanevanwifferen Thanks, is the release built yet for these changes?

nanbingxyz commented:
@lnvestor Some work is still ongoing. It should be finished by around this weekend.
